File: Simulation_NetSwitching/Simulation_NetSwitching/Preference.cpp
#include "Preference.h"
#include "Node.h"
#include "Edge.h"
#include <cmath>

Preference::Preference(Node* node, Edge* edge, bool checkBLE) {
	this->node = node;
	this->edge = edge;
	this->preference = this->getPreference(checkBLE);
}

double Preference::getPreference(bool checkBLE) {
	double min = DBL_MIN;

	this->node->connectType = 2;
	this->node->setProcessSize();
	double p_wifi = 210.0;
	double q_wifi = this->node->dataQueue - this->node->processSize;
	if (q_wifi < 0.0)
		q_wifi = 0.0;

	double v = 3.8;
	min = p_wifi + v * q_wifi;
	this->type = 2;

	if (checkBLE) {
		this->node->connectType = 1;
		this->node->setProcessSize();
		double p_ble = 0.0076 * log(this->node->processSize * 100) + 0.01;
		double q_ble = this->node->dataQueue - this->node->processSize;
		if (q_ble < 0.0)
			q_ble = 0.0;
		double min_ble = p_ble + v * q_ble;

		if (min_ble < min) {
			min = min_ble;
			this->type = 1;
		}
	}
	this->node->connectType = 0;
	this->node->setProcessSize();

	return min;
}

File: Preference.py
import Node
import Edge

class Preference:
    def __init__(self, node, edge, checkBLE):
        self.node = node
        self.edge = edge
        self.type = 0
        self.preference = self.getPreference(checkBLE)       

    def getPreference(self, checkBLE):
        __preference = 0.0
        self.node.connectType = 2
        self.node.getProcessSize()

        p_wifi = 270.0
        q_wifi = self.node.dataQueue *(self.node.inputSize - self.node.processSize)

        if q_wifi < 0.0:
            q_wifi = 0.0
        
        v = 0.0044
        __preference = p_wifi + v * q_wifi
        self.type = 2

        if checkBLE:
            self.node.connectType = 1
            self.node.getProcessSize()
            if 0.6 <= self.node.processSize <= 1.2:
                p_ble = 260.0
            elif self.node.processSize <= 0.6:
                p_ble = 130.0

            q_ble = self.node.dataQueue * (self.node.dataQueue - self.node.processSize)

            if q_ble < 0.0:
                q_ble = 0.0
            min_ble = p_ble + v * q_ble
            
            if min_ble < __preference:
                __preference = min_ble
                self.type = 1

        self.node.connectType = 0
        self.node.getProcessSize()
    
        return __preference